### PR TITLE
Split DisposableWrapper for Native and JVM/JS

### PR DESCRIPTION
**Pull request by arkivanov**

### CODE REVIEW

File: reaktive/src/commonMain/kotlin/com/badoo/reaktive/disposable/DisposableWrapper.kt [hunk 1]
```diff
@@ -1,46 +1,21 @@
...
-import com.badoo.reaktive.utils.atomic.getAndUpdate
-
 /**
  * Thread-safe container of one [Disposable]
  */
-class DisposableWrapper : Disposable {
-
-    private val ref = AtomicReference<Holder?>(Holder(null))
-    override val isDisposed: Boolean get() = ref.value == null
+@Suppress("EmptyDefaultConstructor")
```
Comment (CherryPerry) [inline reaktive/src/commonMain/kotlin/com/badoo/reaktive/disposable/DisposableWrapper.kt]: Quite strange, that we need to suppress it.
https://github.com/arturbosch/detekt/blob/master/detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/empty/EmptyDefaultConstructor.kt#L30

File: reaktive/src/jvmJsCommonMain/kotlin/com/badoo/reaktive/disposable/DisposableWrapper.kt [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    private var disposable: Disposable? = null
+
+    /**
+     * Disposes this [DisposableWrapper] and a stored [Disposable] if any.
+     * Any future [Disposable] will be immediately disposed.
+     */
+    actual override fun dispose() {
+        val disposableToDispose: Disposable?
+
+        synchronized(this) {
```
Comment (CherryPerry) [inline reaktive/src/jvmJsCommonMain/kotlin/com/badoo/reaktive/disposable/DisposableWrapper.kt]: Smaller implementation, but probably less readable.
```kotlin
synchronized(this) {
    _isDisposed = true
    disposable.also { disposable = null }
}?.dispose()
```

File: reaktive/src/jvmJsCommonMain/kotlin/com/badoo/reaktive/disposable/DisposableWrapper.kt [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    /**
+     * Atomically either replaces any existing [Disposable]
+     * with the specified one or disposes it if wrapper is already disposed.
+     * Also disposes any replaced [Disposable].
+     */
+    actual fun set(disposable: Disposable?) {
+        val disposableToDispose: Disposable?
+
+        synchronized(this) {
```
Comment (CherryPerry) [inline reaktive/src/jvmJsCommonMain/kotlin/com/badoo/reaktive/disposable/DisposableWrapper.kt]: Same can be done here.

File: reaktive/src/commonMain/kotlin/com/badoo/reaktive/disposable/DisposableWrapper.kt (hunk 1 above)
Comment (CherryPerry) [reply reaktive/src/commonMain/kotlin/com/badoo/reaktive/disposable/DisposableWrapper.kt]: Can you try `expect class DisposableWrapper expect construction()`?

Comment (arkivanov) [reply reaktive/src/commonMain/kotlin/com/badoo/reaktive/disposable/DisposableWrapper.kt]: `expect` is not allowed for constructors :-(

Comment (CherryPerry) [reply reaktive/src/commonMain/kotlin/com/badoo/reaktive/disposable/DisposableWrapper.kt]: Try just constructor then?

Comment (arkivanov) [reply reaktive/src/commonMain/kotlin/com/badoo/reaktive/disposable/DisposableWrapper.kt]: Still fails

**Pull request merged**